src/pythie_serving/abstract_wrapper.py

```diff
@@ -0,0 +1,99 @@
+
+import grpc
+import numpy as np
+from numpy.typing import NDArray
+
+from .exceptions import PythieServingException
+from .tensorflow_proto.tensorflow_serving.apis import (
+    predict_pb2,
+    prediction_service_pb2_grpc,
+)
+from .tensorflow_proto.tensorflow_serving.config.model_server_config_pb2 import (
+    ModelConfig,
+    ModelServerConfig,
+)
+from .utils import make_ndarray_from_tensor, make_tensor_proto
+
+
+class ModelSpecs(TypedDict):
+    model: Any
+    feature_names: List[str]
+    nb_features: int
+    samples_dtype: Any
+
+
+class AbstractPythieServingPredictionServiceServicer(prediction_service_pb2_grpc.PredictionServiceServicer, abc.ABC):
+    def __init__(self, *, model_server_config: ModelServerConfig):
+        self.logger = logging.getLogger("pythie_serving")
+        self.model_map = {
+            model_config.name: self._create_model_specs(model_config)
+            for model_config in model_server_config.model_config_list.config
+        }
+
+    @abc.abstractmethod
+    def _create_model_specs(self, model_config: ModelConfig) -> ModelSpecs:
+        """Creates the ModelSpecs for a given ModelConfig."""
+        ...
+
+    @abc.abstractmethod
+    def _predict(self, model_specs: ModelSpecs, samples: NDArray) -> NDArray:
+        """How to make a prediction on samples using the model object."""
+        ...
+
+    def Predict(self, request: predict_pb2.PredictRequest, context: grpc.RpcContext) -> predict_pb2.PredictResponse:
+        start = time.time()
+        try:
+            model_name = request.model_spec.name
+            if model_name not in self.model_map:
+                raise PythieServingException(
+                    f"Unknown model: {model_name}. This pythie-serving instance can only "
+                    f'serve one of the following: {",".join(self.model_map.keys())}'
+                )
+            model_specs = self.model_map[model_name]
+            pred = self._predict(model_specs, samples=self._parse_samples(request, model_specs))
+            predict_response = predict_pb2.PredictResponse(
+                model_spec=request.model_spec, outputs={"predictions": make_tensor_proto(pred)}
+            )
+        except Exception as e:
+            self.logger.error(f"Failed to serve because: {e}")
+            raise
+        else:
+            duration = time.time() - start
+            self.logger.info(
+                f"Served model {request.model_spec.name}/{request.model_spec.signature_name}: "
+                f"{len(pred)} predictions in {duration:.2f} seconds ({len(pred) / duration:.2f} pred/sec) "
+            )
+            return predict_response
+
+    @staticmethod
+    def _parse_samples(request: predict_pb2.PredictRequest, model_specs: ModelSpecs) -> NDArray:
+        nb_features = model_specs["nb_features"]
+        request_inputs = request.inputs
+
+        expected_features_names = set(request_inputs)
+        features_names = model_specs["feature_names"]
```

Comment:
Shouldn't it be the contrary ? Expected features are the ones from `model_specs["feature_names"]` and the given one from the request_inputs

Reply:
right!